namematch/generate_must_links.py

```diff
@@ -161,37 +164,34 @@ def get_must_links(self, ml_var_df, uid_vars_list, eid_vars_list, **kw):
                                    f"getting coded as values.")
                     logger.info(uid_counts_high)
                 if uid_counts.max() > 1000:
+                    logger.error("There are uids with more than 1000 records")
                     raise()
 
+            ml_var_nonmissing_ix = ml_var_df[ml_var] != ''
             must_link_df = pd.merge(
-                    ml_var_df[ml_var_df[ml_var] != ''], ml_var_df,
+                    ml_var_df.loc[ml_var_nonmissing_ix],
+                    ml_var_df.loc[ml_var_nonmissing_ix],

```

Comment:
@mmcneill Just want to make sure I understand correctly, I believe this is where the unnecessary merge happened before. It seems like there are 2 changes:

1. Now we only merge the non-missing uid of ml_df_var with itself and do nothing with existing id instead of merge non-missing eid of ml_var_df with itself and merge non-missing uid of ml_var_df with total ml_var_df. 
2. `ml_var_df` now doesn't have columns `has_ml_var` and `file_type` but add a `new_record` column

It does make sense that this should mitigate RAM consumption by ditching some merges and reducing the size of dataframe (ml_df_var). I have two questions regarding this:

1. It seems like it doesn't only reduce RAM for big incremental runs but also regular runs right (if there are eid column(s) provided)?
2. Is the reason why we can ignore eid columns here because eid must links will be taken care in the clustering stage function `def get_initial_clusters`?

Reply:
That's right. 

1. Only incremental runs have existing ID (eid) columns. 
2. That's correct! I just updated the names of several dataframes in the clustering script to hopefully make the logic a little more clear. I was confused myself when I just went to look at this commit's changes in that script.